cache/redisOps.js

```diff
@@ -16,330 +16,388 @@ const redisStore = require('./sampleStore');
-  return getHashPromise(type, name)
-  .then((value) => {
-    redisStore.arrayObjsStringsToJson(
-      value, rsConstant.fieldsToStringify[type]
-    );
-    return Promise.resolve(value);
-  })
-  .catch((err) => Promise.reject(err));
-} // getValue
+  /**
+   * Conditionally add commands to an active batch.
+   *
+   * @param  {Boolean} test - determines whether or not to execute fn
+   * @param  {Function} fn - function that adds commands to the batch
+   * @returns {RedisOps}
+   */
+  if(test, fn) {
+    if (!this.batch) return this;
 
-/**
- * Adds an entry identified by name to the master list of indices identified
- * by "type"
- * @param  {String} type - The type of the master list on which the
- *  set operations are to be performed
- * @param {String} name - Name of the key to be added
- * @returns {Promise|Array} - Promise which resolves to the values returned
- * by the redis command | command array if returnCmd is true.
- */
-function addKey(type, name, returnCmd=false) {
-  const indexName = redisStore.constants.indexKey[type];
-  const nameKey = redisStore.toKey(type, name);
+    if (test) {
+      fn(this);
+    }
 
-  if (returnCmd) {
-    return ['sadd', indexName, nameKey];
+    return this;
   }
 
-  return redisClient.saddAsync(indexName, nameKey)
-  .then((ok) => Promise.resolve(ok))
-  .catch((err) => Promise.reject(err));
-} // addKey
+  /**
+   * Add commands to an active batch by mapping an array.
+   *
+   * @param  {Array} arr - array to map
+   * @param  {Function} fn - function that adds commands to the batch for each array element
+   * @returns {RedisOps}
+   */
+  map(arr, fn) {
+    if (!this.batch) return this;
 
-/**
- * Deletes an entry from the master list of indices identified by "type" and
- * and the corresponding hash identified by "name".
- * @param  {String} type - The type of the master list on which the
- *  set operations are to be performed
- * @param {String} name - Name of the key to be deleted
- * @param {Array} additionalCmds - Additional commands which can be batched
- * @returns {Promise} - which resolves to the values returned by the redis
- * batch command
- */
-function deleteKey(type, name, additionalCmds=[]) {
-  const indexName = redisStore.constants.indexKey[type];
-  const key = redisStore.toKey(type, name);
-  const cmds = [];
+    arr.forEach((element) =>
+      fn(this, element)
+    );
 
-  // remove the entry from the master index of type
-  if (indexName) {
-    cmds.push(['srem', indexName, key]);
+    return this;
   }
 
-  // delete the hash
-  cmds.push(['del', key]);
-  if (additionalCmds && additionalCmds.length > 0) {
-    cmds.push(...additionalCmds);
-  }
+  /**
+   * Add commands to an active batch, making their result available after the
+   * batch is executed.
+   *
+   * @param  {String} key - property name to be set on the result object
+   * @param  {Function} fn - function that adds commands to the batch
+   * @returns {RedisOps}
+   */
+  return(key, fn) {
```

Comment:
"return" seems like a confusing name for a function, surprised it even lets you use it, consider renaming for to avoid any confusion?

Reply:
It's allowed because it's a method, so it will always be called like `batch.return('key', () => ...)`, and never like `return('key', () => ...)`. If you tried to do `const { return } = redisOps` it would fail.

Typical usage is:
```
this.batchCmds()
.doStuff()
.doOtherStuff()
.return('things', (batch) =>
  batch.getThings()
)
.exec()
.then(({ things }) => processThings(things))
```

I tried several different names for this and settled on "return" because it was the only one where it really felt clear what it was doing. It's the same idea as a normal function: you do a bunch of stuff, some of which may produce a value, but only the value from the statement marked with "return" will be used by subsequent code.

With something like:
```
.saveResult('things', (batch) =>
  batch.getThings()
)
```
or
```
.doThen('things', (batch) =>
  batch.getThings()
)
```
the intent seems less clear.

Any other name ideas?

Reply:
OK, go ahead with it as is